sdks/python/apache_beam/runners/worker/statesampler.pyx

```diff
@@ -150,11 +161,31 @@ cdef class StateSampler(object):
           # without the GIL.
           nsecs_ptr = &(<ScopedState>PyList_GET_ITEM(
               self.scoped_states_by_index, self.current_state_index)).nsecs
+          lull_nsecs_ptr = &(<ScopedState>PyList_GET_ITEM(
+              self.scoped_states_by_index, self.current_state_index)).lull_nsecs
+
           nsecs_ptr[0] += elapsed_nsecs
+          lull_nsecs_ptr[0] += elapsed_nsecs
+
+          if lull_nsecs_ptr[0] > self.lull_threshold_ms * MSECS_TO_NSECS:
+            report_lull = True
+            lull_state = self.current_state_index
+
           last_nsecs += elapsed_nsecs
         finally:
           pythread.PyThread_release_lock(self.lock)
 
+        if report_lull:
+          with gil:
```

Comment:
<!--thread_id:cc_143091728_t; commit:95830a027670626ce0f081991d56dac60c8ee8d1; resolved:0-->
<!--section:context-quote-->
> **pabloem** wrote:
> Robert recommended reacquiring the gil and using the Python logging instead of module because that way we'll seamlessly integrate with the portability framework. Also, lulls happen very rarely, when no transition has occurred for 5 minutes, so we rarely incur this cost. What do you think?

<!--section:body-->
What I mean is that we haven't gotten rid of the deadlock possibility here because the two locks could be acquired in either order right now.  We should address that in another PR, and in the meantime, we can log to stderr.

Reply:
To avoid the deadlock, we wait to release the sampler lock in line 176, and acquire the GIL in line 179. That guarantees no deadlocks.